crates/core/src/client/mod.rs

```diff
@@ -57,24 +57,20 @@ where
     proxy_account_class_hash: FieldElement,
 }
 
-impl KakarotClient<JsonRpcClient<HttpTransport>> {
+impl<T: JsonRpcTransport + Send + Sync> KakarotClient<JsonRpcClient<T>> {
     /// Create a new `KakarotClient`.
     ///
     /// # Arguments
     ///
-    /// * `starknet_rpc(&str)` - `StarkNet` RPC
+    /// * `starknet_config(StarknetConfig)` - `StarkNet` configuration.
+    /// * `provider(T)` - `StarkNet` provider.
     ///
     /// # Errors
     ///
-    /// `Err(EthApiError)` if the operation failed.
-    pub fn new(starknet_config: StarknetConfig) -> Result<Self> {
-        let StarknetConfig { starknet_rpc, kakarot_address, proxy_account_class_hash } = starknet_config;
-        let url = Url::parse(&starknet_rpc)?;
-        Ok(Self {
-            starknet_provider: JsonRpcClient::new(HttpTransport::new(url)),
-            kakarot_address,
-            proxy_account_class_hash,
-        })
+    /// `Err(EthApiError<T>)` if the operation failed.
+    pub fn new_with_provider(starknet_config: StarknetConfig, provider: JsonRpcClient<T>) -> Result<Self> {
```

Comment:
imo this can just be called `new`

Reply:
The issue is that we have a specific implementation for `HttpTransport` (which I renamed `new_with_http_transport`). I could remove this specific implementation, which would just require people to pass a provider.